File: apps/android/src/main/kotlin/com/points/android/ui/FocusCounter.kt
@file:OptIn(kotlin.uuid.ExperimentalUuidApi::class)

package com.points.android.ui

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.scale
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.points.core.domain.PointTrend
import com.points.android.ui.theme.Hanken

/**
 * The full-screen Focus counter (design `screens-detail.jsx` → `CounterFS`), reached from the Detail counter
 * row's expand button: the whole screen is the tap target — every tap adds the type's own step with a soft
 * accent ripple where the finger landed and a little pop of the numeral. A round − in the footer corrects
 * overshoots; Done (or back) returns to the Detail screen.
 */
@Composable
fun FocusCounter(
    trend: PointTrend,
    accent: Color,
    onIncrement: () -> Unit,
    onDecrement: () -> Unit,
    onClose: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val haptics = LocalHapticFeedback.current
    val ripples = remember { mutableStateListOf<Ripple>() }
    val pop = remember { Animatable(1f) }

    Column(modifier = modifier.fillMaxSize().background(MaterialTheme.colorScheme.background)) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 8.dp, vertical = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            IconButton(onClick = onClose, modifier = Modifier.semantics { contentDescription = "done" }) {
                StrokeIcon(IconPaths.BACK)
            }
            Text(
                text = "Focus",
                modifier = Modifier.weight(1f),
                textAlign = TextAlign.Center,
                style = TextStyle(fontFamily = Hanken, fontSize = 16.sp, fontWeight = FontWeight.SemiBold),
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            Spacer(Modifier.width(48.dp))
        }

        // the tap field: ripples draw behind the numeral, newest-last
        Column(
            modifier = Modifier
                .weight(1f)
                .fillMaxWidth()
                .drawBehind {
                    ripples.forEach { ripple ->
                        drawCircle(
                            color = accent.copy(alpha = 0.16f * (1f - ripple.progress.value)),
                            radius = 120.dp.toPx() * ripple.progress.value,
                            center = ripple.center,
                        )
                    }
                }
                .pointerInput(Unit) {
                    detectTapGestures(onTap = { offset ->
                        haptics.performHapticFeedback(HapticFeedbackType.LongPress)
                        ripples += Ripple(center = offset)
                        onIncrement()
                    })
                },
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center,
        ) {
            val valueText = formatCount(trend.value)
            // the numeral shrinks as it grows: 150sp up to 3 chars, 104sp to 5, 78sp beyond
            val numSize = when {
                valueText.length <= 3 -> 150.sp
                valueText.length <= 5 -> 104.sp
                else -> 78.sp
            }
            LaunchedEffect(trend.value) {
                pop.snapTo(1.1f)
                pop.animateTo(1f, tween(durationMillis = 320))
            }
            Text(
                text = trend.type.name,
                style = TextStyle(fontFamily = Hanken, fontSize = 17.sp, fontWeight = FontWeight.SemiBold),
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            Text(
                text = valueText,
                modifier = Modifier
                    .padding(vertical = 14.dp)
                    .scale(pop.value)
                    .semantics { contentDescription = "counter value" },
                style = TextStyle(
                    fontFamily = Hanken,
                    fontSize = numSize,
                    fontWeight = FontWeight.Medium,
                    fontFeatureSettings = "tnum",
                ),
                color = MaterialTheme.colorScheme.onSurface,
            )
            val step = trend.type.step.coerceAtLeast(1L)
            Text(
                text = if (step > 1) "Tap anywhere to add $step" else "Tap anywhere to add",
                style = TextStyle(fontFamily = Hanken, fontSize = 14.sp, fontWeight = FontWeight.Medium),
                color = MaterialTheme.colorScheme.outline,
            )
        }

        Row(
            modifier = Modifier.fillMaxWidth().padding(start = 24.dp, end = 24.dp, top = 20.dp, bottom = 30.dp),
            horizontalArrangement = Arrangement.spacedBy(22.dp, Alignment.CenterHorizontally),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                modifier = Modifier
                    .size(64.dp)
                    .clip(CircleShape)
                    .background(MaterialTheme.colorScheme.surfaceContainerHigh)
                    .clickable {
                        haptics.performHapticFeedback(HapticFeedbackType.LongPress)
                        onDecrement()
                    }
                    .semantics { contentDescription = "decrement" },
                contentAlignment = Alignment.Center,
            ) { StrokeIcon(IconPaths.MINUS, size = 26.dp) }
            Text(
                text = "Done",
                modifier = Modifier
                    .clip(CircleShape)
                    .clickable(onClick = onClose)
                    .widthIn(min = 120.dp)
                    .padding(vertical = 14.dp),
                textAlign = TextAlign.Center,
                style = TextStyle(fontFamily = Hanken, fontSize = 15.5.sp, fontWeight = FontWeight.SemiBold),
                color = MaterialTheme.colorScheme.onSurface,
            )
        }
    }

    // each ripple plays its 600ms expansion once, then leaves the list
    ripples.forEach { ripple ->
        LaunchedEffect(ripple) {
            ripple.progress.animateTo(1f, tween(durationMillis = 600))
            ripples.remove(ripple)
        }
    }
}

/** One in-flight tap ripple: where it landed and how far it has expanded. */
private class Ripple(val center: Offset) {
    val progress = Animatable(0f)
}

File: apps/ios/Sources/Detail/FocusCounter.swift
import SwiftUI
import UIKit
import PointsKit

/// The full-screen Focus counter (design `screens-detail.jsx` → `CounterFS`), reached from the Detail counter
/// row's expand button: the whole screen is the tap target — every tap adds the type's own step with a soft
/// accent ripple where the finger landed and a little pop of the numeral. A round − in the footer corrects
/// overshoots; Done (or the back chevron) returns to the Detail screen.
struct FocusCounterView: View {
    let trend: PointTrend
    let accent: Color
    let onIncrement: () -> Void
    let onDecrement: () -> Void
    let onClose: () -> Void

    @State private var ripples: [FocusRipple] = []
    @State private var popScale: CGFloat = 1

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 0) {
                Button(action: onClose) {
                    StrokeIconView(glyph: .back, size: 22)
                        .frame(width: 44, height: 44)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("done")
                Text("Focus")
                    .font(.points(16, .semibold))
                    .foregroundColor(.inkDim)
                    .frame(maxWidth: .infinity)
                Spacer().frame(width: 44)
            }
            .padding(.horizontal, 8)
            .padding(.vertical, 4)

            tapField

            HStack(spacing: 22) {
                Button {
                    UIImpactFeedbackGenerator(style: .medium).impactOccurred()
                    onDecrement()
                } label: {
                    ZStack {
                        Circle().fill(Color.surfaceHigh)
                        StrokeIconView(glyph: .minus, size: 26)
                    }
                    .frame(width: 64, height: 64)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("decrement")

                Button("Done", action: onClose)
                    .font(.points(15.5, .semibold))
                    .foregroundColor(.ink)
                    .frame(minWidth: 120)
                    .padding(.vertical, 14)
            }
            .padding(.top, 20)
            .padding(.bottom, 30)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color.bg.ignoresSafeArea())
    }

    private var tapField: some View {
        let valueText = formatCount(trend.value)
        // the numeral shrinks as it grows: 150pt up to 3 chars, 104pt to 5, 78pt beyond
        let numSize: CGFloat = valueText.count <= 3 ? 150 : (valueText.count <= 5 ? 104 : 78)
        let step = max(trend.type.step, 1)

        return ZStack {
            ForEach(ripples) { ripple in
                FocusRippleView(ripple: ripple, accent: accent)
            }
            VStack(spacing: 0) {
                Text(trend.type.name)
                    .font(.points(17, .semibold))
                    .foregroundColor(.inkDim)
                Text(valueText)
                    .font(.points(numSize, .medium, relativeTo: .largeTitle))
                    .monospacedDigit()
                    .foregroundColor(.ink)
                    .scaleEffect(popScale)
                    .padding(.vertical, 14)
                    .accessibilityLabel("counter value")
                Text(step > 1 ? "Tap anywhere to add \(step)" : "Tap anywhere to add")
                    .font(.points(14, .medium))
                    .foregroundColor(.inkFaint)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .contentShape(Rectangle())
        // DragGesture(minimumDistance: 0) is the iOS-15 way to read the tap's location
        .gesture(
            DragGesture(minimumDistance: 0)
                .onEnded { gesture in tap(at: gesture.location) }
        )
        .onChange(of: trend.value) { _ in
            popScale = 1.1
            withAnimation(.easeOut(duration: 0.32)) { popScale = 1 }
        }
    }

    private func tap(at location: CGPoint) {
        UIImpactFeedbackGenerator(style: .medium).impactOccurred()
        let ripple = FocusRipple(center: location)
        ripples.append(ripple)
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.65) {
            ripples.removeAll { $0.id == ripple.id }
        }
        onIncrement()
    }
}

/// One in-flight tap ripple: where it landed. Its expansion is animated by `FocusRippleView`.
private struct FocusRipple: Identifiable {
    let id = UUID()
    let center: CGPoint
}

private struct FocusRippleView: View {
    let ripple: FocusRipple
    let accent: Color

    @State private var progress: CGFloat = 0

    var body: some View {
        Circle()
            .fill(accent.opacity(0.16 * (1 - progress)))
            .frame(width: 240, height: 240)
            .scaleEffect(progress)
            .position(ripple.center)
            .onAppear {
                withAnimation(.easeOut(duration: 0.6)) { progress = 1 }
            }
            .allowsHitTesting(false)
    }
}
